modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/OAuthDCRMReadTestCase.java

```diff
@@ -0,0 +1,90 @@
+import org.wso2.identity.integration.common.clients.oauth.OauthAdminClient;
+import org.wso2.identity.integration.common.utils.ISIntegrationTest;
+import org.wso2.identity.integration.test.oauth2.dcrm.api.bean.ServiceProviderDataHolder;
+import org.wso2.identity.integration.test.oauth2.dcrm.api.util.OAuthDCRMConstants;
+
+import java.io.*;
+
+import static org.testng.Assert.assertEquals;
+
+/**
+ * OAuth2 DCRM API Read process test case
+ */
+public class OAuthDCRMReadTestCase extends ISIntegrationTest {
+
+    private HttpClient client;
+    private ServiceProviderRegister serviceProviderRegister;
+    private ServiceProviderDataHolder serviceProvider;
+    private OauthAdminClient adminClient;
+    private ApplicationManagementServiceClient appMgtService;
+
+    @BeforeClass(alwaysRun = true)
+    public void testInit() throws Exception {
+        super.init();
+        appMgtService = new ApplicationManagementServiceClient(sessionCookie, backendURL, null);
+        adminClient = new OauthAdminClient(backendURL, sessionCookie);
+        client = HttpClients.createDefault();
+        serviceProviderRegister = new ServiceProviderRegister();
+        serviceProvider = serviceProviderRegister.register(appMgtService, adminClient);
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void atEnd() throws Exception {
+        adminClient.removeOAuthApplicationData(serviceProvider.getClientID());
+        appMgtService.deleteApplication(serviceProvider.getClientName());
+    }
+
+    @Test(alwaysRun = true, description = "Read service provider")
+    public void testReadServiceProvider() throws IOException {
+
+        HttpGet request = new HttpGet(OAuthDCRMConstants.CLIENT_CONFIGURATION_ENDPOINT
+                + serviceProvider.getClientID());
+        request.addHeader(HttpHeaders.AUTHORIZATION, OAuthDCRMConstants.AUTHORIZATION);
+        request.addHeader(HttpHeaders.CONTENT_TYPE, OAuthDCRMConstants.CONTENT_TYPE);
+
+        HttpResponse response = client.execute(request);
+        assertEquals(response.getStatusLine().getStatusCode(), 200, "Service provider request has " +
+                "not returned with successful response");
+    }
+
+    @Test(alwaysRun = true, description = "Read request with an invalid client ID")
+    public void testReadServiceProviderWithInvalidClientID() throws IOException {
+        HttpGet request = new HttpGet(OAuthDCRMConstants.CLIENT_CONFIGURATION_ENDPOINT
+                + OAuthDCRMConstants.INVALID_CLIENT_ID);
+        request.addHeader(HttpHeaders.AUTHORIZATION, OAuthDCRMConstants.AUTHORIZATION);
+        request.addHeader(HttpHeaders.CONTENT_TYPE, OAuthDCRMConstants.CONTENT_TYPE);
+        request.addHeader(HttpHeaders.ACCEPT, OAuthDCRMConstants.ACCEPT);
+
+        HttpResponse response = client.execute(request);
+        assertEquals(response.getStatusLine().getStatusCode(), 500, "Service Provider read request " +
```

Comment:
According to the spec this should have returned a 401. So the implementation may be wrong to respond with 500 in this case. Hence correct the implementation and test case to reflect the spec.

In general, You should not assert a HTTP 500 unless there is a very special case. HTTP 500 can occur due to any reason and asserting a condition based on that is wrong.  This was done in several places, please check and correct.

Reply:
Created https://wso2.org/jira/browse/IDENTITY-6379 to track the issue. Once the implementation is properly fixed, we will fix the test cases.